File: 49.js
// 49. Group Anagrams
/*
# Given an array of strings strs, group the anagrams together. You can return the answer in any order.
# An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once.
*/

var groupAnagrams = function(strs) {
    function alphabetize(s) {
        return s.split('').sort().join('')
    }
    const map = {}
    for (let str of strs) {
        code = alphabetize(str)
        if (map[code]) {
            map[code].push(str)
        } else {
            map[code] = [str]
        }
    }
    const arr = []
    for (let key in map) {
        arr.push(map[key])
    }
    return arr
};

console.log(groupAnagrams(["eat","tea","tan","ate","nat","bat"]))
console.log(groupAnagrams(["cab","tin","pew","duh","may","ill","buy","bar","max","doc"]))

File: 49.py
# 49. Group Anagrams
# Given an array of strings strs, group the anagrams together. You can return the answer in any order.
# An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once.

class Solution:
    def groupAnagrams(self, strs: list[str]) -> list[list[str]]:
        def get_tuple_key(string):
            arr = [0] * 26
            for char in string:
                idx = ord(char) - 97
                arr[idx] += 1
            return tuple(arr)
        
        groups = {}
        for word in strs:
            key = get_tuple_key(word)
            if key in groups:
                groups[key].append(word)
            else:
                groups[key] = [word]
        return groups.values()

s = Solution()
print(s.groupAnagrams(["eat","tea","tan","ate","nat","bat"]))
print(s.groupAnagrams([""]))
print(s.groupAnagrams(["a"]))